src/main/java/com/turn/ttorrent/common/Torrent.java

```diff
@@ -418,6 +418,12 @@ public void save(OutputStream output) throws IOException {
 		return md.digest();
 	}
 
+	public static byte[] hash(ByteBuffer data) throws NoSuchAlgorithmException {
+		MessageDigest md = MessageDigest.getInstance("SHA-1");
+		md.update(data);
```

Comment:
If you look into here : https://github.com/mpetazzoni/ttorrent/pull/35/files#diff-454b627720e29935f25dab9a660670b1L166 you can see a big memory allocation. I don't know how big is it, if it's 64K or 512K - I've implemented it 3.5 years ago - but I'm fairly certain that one torrent piece can be pretty big, and it is much simpler to pass the ByteBuffer directly to the MessageDigest class, instead of allocating a new byte array, copy everything into it, calculate the hash.

 Sorry being not too clear, English is not my native language :(